// ppt-paint-for-animation/webpack.config.js
const path = require('path');
const webpack = require('webpack');

const CleanWebpackPlugin = require('clean-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const HtmlWebpackPlugin = require('html-webpack-plugin');

module.exports = {
    mode: process.env.NODE_ENV || 'production',
    // mode: 'development',
    entry: {
        // qwebchannel:'./public/common/qwebchannel.js',
        // msgutils:'./public/common/msgutils.js',
        index: './js/index.js',
        // pptFotAnimation: './common.js'
    },
    output: {
        path: path.resolve(__dirname, 'public/dist'),
        filename: '[name].js'
    },
    module: {
        rules: [
            { test: /\.js$/, exclude: /node_modules/, loader: "babel-loader" },
            {
                test: /\.scss$/,
                use: [
                    // MiniCssExtractPlugin.loader,
                    'style-loader',
                    'css-loader',
                    'sass-loader'
                ]
            },
            {
                test: /\.(png|jpg|gif)$/,
                use: [
                    {
                        loader: 'url-loader',
                        options: {
                            limit: 8192
                        }
                    }
                ]
            }
        ]
    },
    plugins: [
        new MiniCssExtractPlugin({
            filename: '[name].css',
            chunkFilename: '[id].css'
        }),
        new CleanWebpackPlugin(['public/dist']),
        // new HtmlWebpackPlugin({
        //     template: './public/start.html',
        //     chunks: ['qwebchannel','msgutils','index'],
        //     filename:'index.html'
        // })
    ],
    watch: true
};
