// ppt-paint-for-animation/webpack.config.pro.js
const path = require('path');
const webpack = require('webpack');

const HtmlWebpackPlugin = require('html-webpack-plugin');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const CleanWebpackPlugin = require('clean-webpack-plugin');

module.exports = {
    mode: process.env.NODE_ENV || 'production',
    entry: './js/index.js',
    output: {
        path: path.resolve(__dirname, 'dist'),
        // filename: 'index.[chunkhash].js',
        filename: 'pptForAnimation.js'
    },
    module: {
        rules: [
            { test: /\.js$/, exclude: /node_modules/, loader: "babel-loader" },
            {
                test: /\.scss$/,
                use: [
                    // MiniCssExtractPlugin.loader,
                    'style-loader',
                    'css-loader',
                    'sass-loader'
                ]
            }
        ]
    },
    plugins: [
        new MiniCssExtractPlugin({
            filename: '[name].[chunkhash].css',
            chunkFilename: '[id].css'
        }),
        new CleanWebpackPlugin(['dist']),
        new HtmlWebpackPlugin({
            template: './public/start.html',
            chunks: ['qwebchannel', 'index'],
            filename: 'pptForAnimation.ftl'
        }),
        new HtmlWebpackPlugin({
            template: './public/startForMobile.html',
            chunks: ['qwebchannel', 'index'],
            filename: 'pptForMobile.ftl'
        })
    ]
};
